devito/types/basic.py

```diff
@@ -426,8 +442,114 @@ def __dtype_setup__(cls, **kwargs):
         return kwargs.get('dtype', np.float32)
 
 
-class AbstractFunction(sympy.Function, Basic, Cached, Pickable, Evaluable):
+class AbstractTensor(sympy.ImmutableDenseMatrix, Basic, Cached, Pickable, Evaluable):
+    """
+    Base class for vector and tensor valued functions. It inherits from and
+    mimicks the behavior of a sympy.ImmutableDenseMatrix.
+
+
+    The sub-hierachy is as follows
+
+                         AbstractTensor
+                                |
+                          TensorFunction
+                                |
+                 ---------------------------------
+                 |                               |
+          VectorFunction                 TensorTimeFunction
+                        \-------\                |
+                                 \------- VectorimeFunction
+
+    There are four relevant AbstractTensor sub-types: ::
+
```

Comment:
Are the definitions below deliberately opposite? Am I missing something big here?

Reply:
?

Reply:
It is fixed now. Case resolved.